resources/sql/upgrade-229.sql

```diff
@@ -46,4 +46,76 @@ update usergroupaccess set access = 'r-r-----' where access = 'r-------' and use
 
 -- set all tracked entity attributes searchable
 
-update program_attributes set searchable = true;
\ No newline at end of file
+update program_attributes set searchable = true;
+
+	
+-- migrate userrole.dataSets and userrole.programs to usergroup and apply data sharing
+-- after executing below scripts run command "select migrateRoleToUserGroup();" to migrate data. 
+
+CREATE OR REPLACE FUNCTION uid()
+RETURNS text AS $$
+  SELECT substring('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ' 
+    FROM (random()*51)::int +1 for 1) || 
+    array_to_string(ARRAY(SELECT substring('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789' 
+       FROM (random()*61)::int + 1 FOR 1) 
+   FROM generate_series(1,10)), '') ;
+$$ LANGUAGE sql;
+
+create or replace function migrateRoleToUserGroup() returns void as $$
+declare role RECORD;
+declare curUserGroupId int;
+declare roleDataset RECORD;
+declare curGroupAccessId int;
+declare roleProgram RECORD;
+begin
+for role in select * from userrole
+loop 
+	if exists ( select 1 from userroledataset rds where rds.userroleid = role.userroleid )
+	then 
+		insert into usergroup (usergroupid, name, uid, code, lastupdated, created, userid, publicaccess, lastupdatedby)
+		values ( nextval('hibernate_sequence'::regclass), '_DATASET_' || role.name, uid(), null,null,now(), role.userid, 'rw------',null )
+		returning usergroup.usergroupid into curUserGroupId;
+		for roleDataset in select * from userroledataset
+		loop 
+			insert into usergroupaccess ( usergroupaccessid, access, usergroupid )
+			values ( nextval('hibernate_sequence'::regclass), 'rwrw----', curUserGroupId )
```

Comment:
We should probably not give metadata write. Only metadata read is required by the data read and write.